src/commands/get_sale_ticket.rs

```diff
@@ -0,0 +1,58 @@
+use candid::Encode;
+use clap::Parser;
+
+use crate::{
+    commands::transfer::HexSubaccount,
+    lib::{
+        signing::{sign_ingress_with_request_status_query, IngressWithRequestId},
+        AnyhowResult, TargetCanister,
+    },
+    SnsCanisterIds,
+};
+
+/// Get the sale ticket of the caller. If there is no open ticket yet, create a new ticket with specified arguments.
+#[derive(Parser)]
+pub struct GetSaleTicketOpts {
+    /// The amount of ICP tokens in e8s.
+    #[clap(long)]
+    amount_icp_e8s: u64,
+
+    /// The subaccount of the account to create sale ticket. For example: e000d80101
+    #[clap(long)]
+    subaccount: Option<HexSubaccount>,
+}
+
+// TODO: SDK-954 - use ic_sns_swap when it is available
+#[derive(candid::CandidType, candid::Deserialize)]
+struct NewSaleTicketRequest {
+    amount_icp_e8s: u64,
+    subaccount: Option<[u8; 32]>,
+}
+
+#[derive(candid::CandidType, candid::Deserialize)]
+struct GetOpenTicketArg {}
+
+pub fn exec(
+    pem: &str,
+    sns_canister_ids: &SnsCanisterIds,
+    opts: GetSaleTicketOpts,
+) -> AnyhowResult<Vec<IngressWithRequestId>> {
+    let req1 = sign_ingress_with_request_status_query(
+        pem,
+        "get_open_ticket",
+        Encode!(&GetOpenTicketArg {})?,
+        TargetCanister::Swap(sns_canister_ids.swap_canister_id.get().0),
+    )?;
+
+    let message = NewSaleTicketRequest {
+        amount_icp_e8s: opts.amount_icp_e8s,
+        subaccount: opts.subaccount.map(|sub| sub.0),
+    };
+    let req2 = sign_ingress_with_request_status_query(
+        pem,
+        "new_sale_ticket",
+        Encode!(&message)?,
+        TargetCanister::Swap(sns_canister_ids.swap_canister_id.get().0),
+    )?;
```

Comment:
It seems like `sns-quill` makes this conditional message execution pretty difficult. I.e. if `get_open_ticket` does return a ticket id, then the subsequent call to `new_sale_ticket` will fail. This may be a confusing experience for users so I've asked Mario if he can introduce a third API to `get_or_create`. You can see the slack thread here https://dfinity.slack.com/archives/C03H6QEPW5D/p1675465230570649?thread_ts=1675169129.448279&cid=C03H6QEPW5D. If that change is impossible, I only see two ways of proceeding:

1. Continue with what is already implemented. The user will have to parse the results.
2. Separate these into two separate commands and have user understand the sale payment flow.

I think (1) is the best option right now but less than ideal. Lets see how he responds on Monday.